pilot/pkg/secrets/kube/secrets.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	informersv1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pilot/pkg/secrets"
+	"istio.io/pkg/log"
+)
+
+const (
+	// The ID/name for the certificate chain in kubernetes generic secret.
+	GenericScrtCert = "cert"
+	// The ID/name for the private key in kubernetes generic secret.
+	GenericScrtKey = "key"
+	// The ID/name for the CA certificate in kubernetes generic secret.
+	GenericScrtCaCert = "cacert"
+
+	// The ID/name for the certificate chain in kubernetes tls secret.
+	TLSSecretCert = "tls.crt"
+	// The ID/name for the k8sKey in kubernetes tls secret.
+	TLSSecretKey = "tls.key"
+	// The ID/name for the CA certificate in kubernetes tls secret
+	TLSSecretCaCert = "ca.crt"
+
+	// GatewaySdsCaSuffix is the suffix of the sds resource name for root CA. All resource
+	// names for gateway root certs end with "-cacert".
+	GatewaySdsCaSuffix = "-cacert"
+)
+
+type SecretsController struct {
+	secrets informersv1.SecretInformer
+}
+
+var _ secrets.Controller = &SecretsController{}
+
+func NewSecretsController(informer informersv1.SecretInformer) *SecretsController {
+	// Informer is lazy loaded, load it now
+	_ = informer.Informer()
+	return &SecretsController{informer}
+}
+
+func (s *SecretsController) GetKeyAndCert(name, namespace string) (key []byte, cert []byte) {
+	k8sSecret, err := s.secrets.Lister().Secrets(namespace).Get(name)
+	if err != nil {
+		return nil, nil
+	}
+
+	return extractKeyAndCert(k8sSecret)
+}
+
+func (s *SecretsController) GetCaCert(name, namespace string) (cert []byte) {
+	strippedName := strings.TrimSuffix(name, GatewaySdsCaSuffix)
+	k8sSecret, err := s.secrets.Lister().Secrets(namespace).Get(strippedName)
+	var rootCert []byte
+	if err != nil {
+		// Could not fetch cert, look for legacy secret with -cacert suffix
```

Comment:
Where is the legacy one used?

Reply:
See supported formats: https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway